Fixing API compat issues between ref/src for attributes on properties

ericstj opened the pull request:
API compat had a regression about a year ago where it stopped checking attributes on members other than methods.  https://github.com/dotnet/arcade/pull/9168 fixes that.  This PR brings us back up to date.

Marking this breaking because one member was missing obsolete in the ref assembly.

ericstj committed:
Fixing API compat issues between ref/src for attributes on properties

ericstj commented on src/libraries/System.Data.Common/ref/System.Data.Common.cs:
```diff
@@ -1577,6 +1577,7 @@ public enum SchemaType
     public enum SerializationFormat
     {
         Xml = 0,
+        [System.ObsoleteAttribute("SerializationFormat.Binary is obsolete and should not be used. See https://aka.ms/serializationformat-binary-obsolete for more information.", DiagnosticId = "SYSLIB0038")]
```
This is a breaking change: it was already made in the src but never shipped in references.

roji replied:
Yeah, we did this recently in #39289 as part of obsoleting (and eventually removing) all uses of BinaryFormatter, for 7.0 (so hasn't shipped yet). It's indeed being treated and documented as a breaking change.

rzikm reviewed (approved):
System.Net.* changes LGTM.

stephentoub reviewed (approved):
Other than the mono build break, LGTM.

ericstj committed:
Fix some build breaks due to obsoleted API

terrajobst commented:
I've used the patch file to get to a quick summary of all new attributes:

<details>
<summary>All attribute additions</summary>

```C#
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.PublicFields)]
[EditorBrowsable(EditorBrowsableState.Advanced)]
[EditorBrowsable(EditorBrowsableState.Never)]
[Obsolete("Add(String parameterName, Object value) has been deprecated. Use AddWithValue(String parameterName, Object value) instead.")]
[Obsolete("Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead.", DiagnosticId = "SYSLIB0012", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
[Obsolete("Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location.", DiagnosticId = "SYSLIB0012", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
[Obsolete("IPAddress.Address is address family dependent and has been deprecated. Use IPAddress.Equals to perform comparisons instead.")]
[Obsolete("IsolatedStorageFile.CurrentSize has been deprecated because it is not CLS Compliant. To get the current size use IsolatedStorageFile.UsedSize instead.")]
[Obsolete("IsolatedStorageFile.MaximumSize has been deprecated because it is not CLS Compliant. To get the maximum size use IsolatedStorageFile.Quota instead.")]
[Obsolete("Marshalling arbitrary types may be unavailable in future releases. Specify the type you wish to marshal as.")]
[Obsolete("Marshalling as AnsiBStr may be unavailable in future releases.")]
[Obsolete("Marshalling as Currency may be unavailable in future releases.")]
[Obsolete("Marshalling as TBstr may be unavailable in future releases.")]
[Obsolete("Marshalling as VBByRefString may be unavailable in future releases.")]
[Obsolete("OffsetToStringData has been deprecated. Use string.GetPinnableReference() instead.")]
[Obsolete("PerformanceCounter.DefaultFileMappingSize has been deprecated and is not used. Use machine.config or an application configuration file to set the size of the PerformanceCounter file mapping instead.")]
[Obsolete("PerformanceCounterManager has been deprecated. Use the PerformanceCounters through the System.Diagnostics.PerformanceCounter class instead.")]
[Obsolete("Process.NonpagedSystemMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.NonpagedSystemMemorySize64 instead.")]
[Obsolete("Process.PagedMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PagedMemorySize64 instead.")]
[Obsolete("Process.PagedSystemMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PagedSystemMemorySize64 instead.")]
[Obsolete("Process.PeakPagedMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PeakPagedMemorySize64 instead.")]
[Obsolete("Process.PeakVirtualMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PeakVirtualMemorySize64 instead.")]
[Obsolete("Process.PeakWorkingSet has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PeakWorkingSet64 instead.")]
[Obsolete("Process.PrivateMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.PrivateMemorySize64 instead.")]
[Obsolete("Process.VirtualMemorySize has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.VirtualMemorySize64 instead.")]
[Obsolete("Process.WorkingSet has been deprecated because the type of the property can't represent all valid results. Use System.Diagnostics.Process.WorkingSet64 instead.")]
[Obsolete("ReflectionPermissionAttribute.ReflectionEmit has been deprecated and is not supported.")]
[Obsolete("ReflectionPermissionAttribute.TypeInformation has been deprecated and is not supported.")]
[Obsolete("ReflectionPermissionFlag.AllFlags has been deprecated. Use PermissionState.Unrestricted to get full access.")]
[Obsolete("ReflectionPermissionFlag.ReflectionEmit  has been deprecated and is not supported.")]
[Obsolete("ReflectionPermissionFlag.TypeInformation has been deprecated and is not supported.")]
[Obsolete("RegistryPermissionAttribute.Add has been deprecated. Use ViewAndModify instead.")]
[Obsolete("SecurityProtocolType.Ssl3 has been deprecated and is not supported.")]
[Obsolete("SerializationFormat.Binary is obsolete and should not be used. See https://aka.ms/serializationformat-binary-obsolete for more information.", DiagnosticId = "SYSLIB0038")]
[Obsolete("StandardEventKeywords.CorrelationHint has an incorrect value and has been deprecated. Use CorrelationHint2 instead.")]
[Obsolete("The ApartmentState property has been deprecated. Use GetApartmentState, SetApartmentState or TrySetApartmentState instead.")]
[Obsolete("vbNewLine has been deprecated. For a carriage return and line feed, use vbCrLf. For the current platform's newline, use System.Environment.NewLine."]
[Obsolete("WaitHandle.Handle has been deprecated. Use the SafeWaitHandle property instead.")]
[RequiresAssemblyFiles("Returns <Unknown> for modules with no file path")]
[RequiresAssemblyFiles("This member throws an exception for assemblies embedded in a single-file app")]
[SupportedOSPlatform("maccatalyst")]
[UnsupportedOSPlatform("freebsd")]
[UnsupportedOSPlatform("illumos")]
[UnsupportedOSPlatform("ios")]
[UnsupportedOSPlatform("osx")]
[UnsupportedOSPlatform("solaris")]
[UnsupportedOSPlatform("tvos")]
```

</details>

These are the types:

* `DynamicallyAccessedMembers`, `RequiresAssemblyFiles`. Linker related. I wouldn't consider them source breaking as they make it easier to trouble should runtime issues when trimming your app.
* `EditorBrowsable`. Not source breaking.
* `Obsolete`. Source breaking, but we have said in the past we're OK with that.
* `SupportedOSPlatform`, `UnsupportedOSPlatform`. Platform compat annotations. I wouldn't consider them source breaking as they make it easier to trouble should runtime issues when running on a different operating system.

eerhardt commented on src/libraries/System.Diagnostics.Process/ref/System.Diagnostics.Process.cs:
```diff
@@ -69,6 +69,8 @@ public partial class Process : System.ComponentModel.Component, System.IDisposab
         public int PrivateMemorySize { get { throw null; } }
         public long PrivateMemorySize64 { get { throw null; } }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("ios")]
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("tvos")]
```
How come all the other APIs (including `ProcessThread.PrivilegedProcessorTime`) has `SupportedOSPlatformAttribute("maccatalyst")`, but this one doesn't?

terrajobst replied:
/cc @buyaa-n @dotnet/area-system-diagnostics-process

ericstj replied:
I'm just making them consistent with https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.iOS.cs#L32-L38

I opened https://github.com/dotnet/runtime/issues/68729

ericstj committed:
Fix remaining attribute inconsistencies

ericstj committed:
Update API compat baselines for the shared framework.

ericstj committed:
Temporarily suppress ApiCompat baseline validation

This will allow us to ensure the codebase is ready to ingest the arcade
fix to ApiCompat.

ericstj commented on src/libraries/apicompat/ApiCompat.proj:
```diff
@@ -7,6 +7,8 @@
     <!-- Target reference assemblies instead of implementation assemblies. -->
     <CompileUsingReferenceAssemblies>true</CompileUsingReferenceAssemblies>
     <TrimOutPrivateAssembliesFromReferencePath>true</TrimOutPrivateAssembliesFromReferencePath>
+    <!-- Temporarily disable baseline validation so that we can add baselines for the upcoming API compat fix https://github.com/dotnet/arcade/pull/9168 -->
```
This flag makes sure that our baseline files are *minimal*.  That fails right now because the current APICompat doesn't generate any of these added errors.  I'll disable that checking until we get the API compat update.

ViktorHofer replied:
Can you please add a tracking issue for that? In the past we often forgot to turn such knobs on again.

bartonjs reviewed (approved):
LGTM, except for the System.Data.Common.Tests build failures.

ericstj committed:
Fix usage of obsolete API in Data.Common tests

ericstj commented:
Doc was created in https://github.com/dotnet/docs/issues/28726 and is published as https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/7.0/serializationformat-binary
